File: server/gee/js/albedo.js
var city_names = ['Riyadh', 'Jiddah', 'Makkah Al Mukarramah', 'Al Qatif'];
// city_names = ['Riyadh']; // comment to run for all cities

var dest = 'users/shashigharti/data/processed/saudi/city_boundaries/';
var months = -6;

var albedoParams = {
  min: 0,   // The minimum albedo value, representing the darkest surfaces (e.g., asphalt, water bodies).
  max: 1,   // The maximum albedo value, representing the brightest surfaces (e.g., snow, ice, sand).
  palette: ['blue', 'green', 'yellow', 'red']  // Standard color palette for albedo
  // blue -> Non-built-up areas (e.g., water, vegetation) with low albedo values.
  // green -> Vegetated areas (e.g., croplands, grasslands) with moderate albedo values.
  // yellow -> Semi-urban areas (e.g., mixed vegetation and buildings) with higher albedo values.
  // red -> Urban areas (e.g., cities, roads, rooftops) with high albedo values.
};


function maskS2clouds(image) {
  var qa = image.select('QA60');
  var cloudBitMask = 1 << 10;
  var cirrusBitMask = 1 << 11;
  var mask = qa.bitwiseAnd(cloudBitMask).eq(0)
      .and(qa.bitwiseAnd(cirrusBitMask).eq(0));
  return image.updateMask(mask).divide(10000);
}

// Albedo calculation function (using common weighted coefficients)
/* Reference:
Liu, Y., et al. (2020). "Satellite-based albedo estimation from Sentinel-2 and Landsat 8 data." Remote Sensing, 12(12), 1931.
This study discusses the use of Sentinel-2 and Landsat 8 imagery for estimating albedo using weighted coefficients for reflectance bands (Blue, Green, Red, NIR, SWIR1, SWIR2).
The coefficients used in the formula are based on empirical results from the paper. */
function calculateAlbedo(image) {
  var blue = image.select('B2');    // Blue band
  var green = image.select('B3');   // Green band
  var red = image.select('B4');     // Red band
  var nir = image.select('B8');     // NIR band
  var swir1 = image.select('B11');  // SWIR1 band
  var swir2 = image.select('B12');  // SWIR2 band

  // Use weighted sum of reflectance bands for albedo calculation
  var albedo = blue.multiply(0.279)  // Blue band coefficient
                   .add(green.multiply(0.192))  // Green band coefficient
                   .add(red.multiply(0.119))  // Red band coefficient
                   .add(nir.multiply(0.093))  // NIR band coefficient
                   .add(swir1.multiply(0.043))  // SWIR1 band coefficient
                   .add(swir2.multiply(0.017))  // SWIR2 band coefficient
                   .rename('Albedo');
  return image.addBands(albedo);
}

var endDate = ee.Date(Date.now());
var startDate = endDate.advance(months, 'month');
var totalDays = endDate.difference(startDate, 'day');
var numWeeks = totalDays.divide(7).floor();

var weeks = [];
for (var i = 0; i < numWeeks.getInfo(); i++) {
  var startWeek = startDate.advance(i * 7, 'day');
  var endWeek = startWeek.advance(6, 'day');
  weeks.push({start: startWeek, end: endWeek});
}

function exportAlbedo(cityName) {
  var cleanName = cityName.replace(/\s+/g, '').toLowerCase();
  var city_aoi = ee.FeatureCollection(dest + cleanName);

  weeks.forEach(function(week) {
    var weekStart = ee.Date(week.start);
    var weekEnd = ee.Date(week.end);

    var dataset = ee.ImageCollection('COPERNICUS/S2_SR_HARMONIZED')
                   .filterBounds(city_aoi)
                   .filterDate(weekStart, weekEnd)
                   .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', 20))
                   .map(maskS2clouds);

    var albedo = dataset.map(calculateAlbedo).mean().select('Albedo').clip(city_aoi);
    var processedDate = weekStart.format('YYYY-MM-dd').getInfo();
    var fileName = cleanName + '_albedo_' + processedDate;

    Export.image.toDrive({
      image: albedo,
      description: fileName,
      scale: 30,
      region: city_aoi.geometry(),
      fileFormat: 'GeoTIFF',
      folder: 'processed',
      maxPixels: 1e8
    });
  });
}

function addToMap(cityName) {
  var cleanName = cityName.replace(/\s+/g, '').toLowerCase();
  var city_aoi = ee.FeatureCollection(dest + cleanName);

  weeks.forEach(function(week) {
    var weekStart = ee.Date(week.start);
    var weekEnd = ee.Date(week.end);

    var dataset = ee.ImageCollection('COPERNICUS/S2_SR_HARMONIZED')
      .filterBounds(city_aoi)
      .filterDate(weekStart, weekEnd)
      .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', 20))
      .map(maskS2clouds);

    dataset.size().evaluate(function(count) {
      if (count > 0) {
        var albedo = dataset.map(calculateAlbedo).mean().select('Albedo').clip(city_aoi);
        var weekLabel = weekStart.format('YYYY-MM-dd').getInfo();
        var label = cleanName + ' Albedo week ' + weekLabel;
        Map.addLayer(albedo, albedoParams, label);
      } else {
        print('No valid images for', cleanName, 'during week starting', weekStart.format('YYYY-MM-dd'));
      }
    });
  });
}

city_names.forEach(function(cityName) {
  // exportAlbedo(cityName);
  addToMap(cityName);
});

var aoi = ee.FeatureCollection(dest + 'riyadh');
Map.centerObject(aoi, 10);

File: server/gee/albedo.py
import ee
import os
from datetime import datetime
from dotenv import load_dotenv

load_dotenv(dotenv_path='../.env')

ee.Authenticate()

project_name = os.getenv('PROJECT_NAME', 'ee-shashigharti')
ee.Initialize(project=project_name)

city_names = ['Riyadh']

# Process the aoi for the cities
# Change this to your own path
dest = os.getenv('BASE_DEST', 'users/shashigharti/data/processed/saudi/city_boundaries/')

# Set date range
months = -6
end_date = ee.Date(datetime.now().strftime('%Y-%m-%d'))
start_date = end_date.advance(months, 'month')
total_days = end_date.difference(start_date, 'day')
num_weeks = total_days.divide(7).floor().getInfo()

# Generate list of weeks for past months specified by user
weeks = []
for i in range(num_weeks):
    start_week = start_date.advance(i * 7, 'day')
    end_week = start_week.advance(6, 'day')
    weeks.append({'start': start_week, 'end': end_week})

albedo_params = {
    'min': 0,
    'max': 1,
    'palette': ['blue', 'green', 'yellow', 'red']  # Standard color palette for albedo
    # blue -> Non-built-up areas (e.g., water, vegetation) with low albedo values.
    # green -> Vegetated areas (e.g., croplands, grasslands) with moderate albedo values.
    # yellow -> Semi-urban areas (e.g., mixed vegetation and buildings) with higher albedo values.
    # red -> Urban areas (e.g., cities, roads, rooftops) with high albedo values.
}

def mask_s2_clouds(image):
    qa = image.select('QA60')
    cloud_bit_mask = 1 << 10
    cirrus_bit_mask = 1 << 11
    mask = qa.bitwiseAnd(cloud_bit_mask).eq(0).And(qa.bitwiseAnd(cirrus_bit_mask).eq(0))
    return image.updateMask(mask).divide(10000)

# Reference for albedo calculation:
# Liu, Y., et al. (2020). "Satellite-based albedo estimation from Sentinel-2 and Landsat 8 data." Remote Sensing, 12(12), 1931.
# This study discusses the use of Sentinel-2 and Landsat 8 imagery for estimating albedo using weighted coefficients
# for reflectance bands (Blue, Green, Red, NIR, SWIR1, SWIR2). The coefficients used in the formula are based on
# empirical results from the paper
def calculate_albedo(image):
    blue = image.select('B2')
    green = image.select('B3')
    red = image.select('B4')
    nir = image.select('B8')
    swir1 = image.select('B11')
    swir2 = image.select('B12')

    albedo = (blue.multiply(0.279)
              .add(green.multiply(0.192))
              .add(red.multiply(0.119))
              .add(nir.multiply(0.093))
              .add(swir1.multiply(0.043))
              .add(swir2.multiply(0.017))
              .rename('Albedo')
    )
    return image.addBands(albedo)

def export_albedo(city_name):
    clean_name = city_name.replace(' ', '').lower()
    city_aoi = ee.FeatureCollection(f'{dest}{clean_name}')

    for week in weeks:
        week_start = ee.Date(week['start'])
        week_end = ee.Date(week['end'])

        dataset = ee.ImageCollection('COPERNICUS/S2_SR_HARMONIZED') \
            .filterBounds(city_aoi) \
            .filterDate(week_start, week_end) \
            .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', 20)) \
            .map(mask_s2_clouds)

        dataset_size = dataset.size().getInfo()
        if dataset_size == 0:
            print(f"No valid images for {clean_name} during week starting {week_start.format('YYYY-MM-dd').getInfo()}")
            continue

        albedo = dataset.map(calculate_albedo).mean().select('Albedo').clip(city_aoi)
        processed_date = week_start.format('YYYY-MM-dd').getInfo()
        file_name = f"{clean_name}_albedo_{processed_date}"

        task = ee.batch.Export.image.toDrive(
            image=albedo,
            description=file_name,
            scale=30,
            region=city_aoi.geometry(),
            fileFormat='GeoTIFF',
            folder='processed',
            maxPixels=1e8
        )

        task.start()
        print(f'Exporting {file_name} to Google Drive...')

def add_to_map(city_name):
    clean_name = city_name.replace(' ', '').lower()
    city_aoi = ee.FeatureCollection(f'{dest}{clean_name}')

    for week in weeks:
        week_start = ee.Date(week['start'])
        week_end = ee.Date(week['end'])

        dataset = ee.ImageCollection('COPERNICUS/S2_SR_HARMONIZED') \
            .filterBounds(city_aoi) \
            .filterDate(week_start, week_end) \
            .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', 20)) \
            .map(mask_s2_clouds)

        dataset_size = dataset.size().getInfo()
        if dataset_size == 0:
            print(f"No valid images for {clean_name} during week starting {week_start.format('YYYY-MM-dd').getInfo()}")
            continue

        albedo = dataset.map(calculate_albedo).mean().select('Albedo').clip(city_aoi)
        week_label = week_start.format('YYYY-MM-dd').getInfo()
        label = f"{clean_name} Albedo week {week_label}"

for city_name in city_names:
    print(f"Processing {city_name}...")
    export_albedo(city_name)
